/*
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

syntax = "proto3";
import "google/protobuf/any.proto";
package NES;

option java_package = "stream.nebula.protobuf";
option java_multiple_files = true;

/*
The serializable wrapper definition for data type objects
 */
message SerializableDataType{
  // Enum do identify the data type
  enum Type{
    UNDEFINED = 0;
    INTEGER = 1;
    FLOAT = 2;
    CHAR = 3;
    ARRAY = 4;
    BOOLEAN = 5;
    TEXT = 6;
  }

  /**
  Details for array data types
   */
  message ArrayDetails{
    SerializableDataType componentType = 1;
    uint64 dimensions = 2;
  }

  /**
  Details for integer data types
   */
  message IntegerDetails{
    uint64 bits = 1;
    int64 upperBound = 2;
    int64 lowerBound = 3;
  }

  /**
 Details for Float data types
  */
  message FloatDetails{
    uint64 bits = 1;
    double upperBound = 2;
    double lowerBound = 3;
  }

  /**
  Details for char data types
   */
  message CharDetails{
    uint64 dimensions = 2;
  }

  Type type = 1;
  // field for custom type properties
  google.protobuf.Any details = 2;

}

/**
 The serializable wrapper definition for data value objects
 */
message SerializableDataValue {

  /**
   An basic data type value
   */
  message BasicValue {
    SerializableDataType type = 1;
    string value = 2;
  }

  /**
   An array value
   */
  message ArrayValue {
    SerializableDataType type = 1;
    repeated string values = 2;
  }

  google.protobuf.Any value = 1;
}